### PR TITLE
Use requirements.txt as lockfile

### PR DESCRIPTION
**Pull request by wuan**
The basic requirement is contained in `requirements.in`. `requirements.txt` can be updated by using `pip-compile` from the  `pip-tools` package.

### CODE REVIEW

File: mailbox/requirements.in [hunk 1]
```diff
@@ -0,0 +1 @@
+magic-wormhole-mailbox-server -e git+https://github.com/magic-wormhole/magic-wormhole-mailbox-server/tarball/4b358859ba80de37c3dc0a5f67ec36909fd48234#egg=magic-wormhole-mailbox-server
```
Comment (meejah) [inline mailbox/requirements.in:1]: I think these github hashes (of magically-produced tarballs) are not stable...?

**Comment by meejah** [comment]
It seems to me that overall pinned requirements would best come from upstream, as is common in lots of other Python command-line application (e.g. `magic-folder`) and ideally come with hashes as well.

File: mailbox/requirements.in (hunk 1 above)
Comment (wuan) [reply mailbox/requirements.in:1]: That hash is the commit hash from the repo:

https://github.com/magic-wormhole/magic-wormhole-mailbox-server/commit/4b358859ba80de37c3dc0a5f67ec36909fd48234

Pretty stable, you can also use a tag (but this is less stable).

Please note that the source URL did not change with this PR, we already used it before.

**Comment by wuan** [comment]
> It seems to me that overall pinned requirements would best come from upstream, as is common in lots of other Python command-line application (e.g. `magic-folder`) and ideally come with hashes as well.

Projects usually allow ranges of versions in order to work together with other dependencies, for example see [magic-wormhole-mailbox-server/setup.py Line 33](https://github.com/magic-wormhole/magic-wormhole-mailbox-server/blob/39672ae95a2635ba9daaba62f483aa75bfd80a22/setup.py#L33):
```
      install_requires=[
          "attrs >= 16.3.0", # 16.3.0 adds __attrs_post_init__
          "twisted[tls] >= 17.5.0",
          "autobahn[twisted] >= 0.14.1",
      ],
```
That is the reason why we lock the dependency versions here in order to have a better reproducibility.

File: mailbox/requirements.in (hunk 1 above)
Comment (meejah) [reply mailbox/requirements.in:1]: Maybe it's a different thing, but there was some issue with using GitHub links (e.g. from the "Download" button, approximately). Maybe it was only for ZIPs?

**Comment by meejah** [comment]
The recommended approach is for _libraries_ to  be open-ended, but for _programs_ to be way more specific (ideally, locking to precise versions). The latter is what e.g. magic-folder does: https://github.com/LeastAuthority/magic-folder/blob/main/requirements/base.txt

**Comment by meejah** [comment]
(I suspect the `setup.py` stuff is basically just "leftover" from before the 3-way split...? Then, it did make sense to be open-ended because part of the repo was 'a library')

**Comment by meejah** [comment]
> That is the reason why we lock the dependency versions here in order to have a better reproducibility.

Yeah, what I'm suggesting is that it could be better to put that effort upstream, since it makes sense to lock them there too. (That is, because it's "a program" and not "a library").

**Comment by wuan** [comment]
> Yeah, what I'm suggesting is that it could be better to put that effort upstream, since it makes sense to lock them there too. (That is, because it's "a program" and not "a library").

The just open an issue upstream, this topic needs to be discussed there and reference this repo.

**Comment by meejah** [comment]
Sure, please open a ticket -- I don't know your requirements nor what discussion you believe needs to happen?

**Comment by wuan** [comment]
> Sure, please open a ticket -- I don't know your requirements nor what discussion you believe needs to happen?

You had the idea to do this, there are no requirements outside of what we were talking about. So just go ahead.

File: mailbox/requirements.in (hunk 1 above)
Comment (meejah) [reply mailbox/requirements.in:1]: Right, okay that was it -- that the bits can change (so if you have a workflow that checks those, it'll fail -- but only "somtimes" e.g. when GitHub nukes a cache or something?)

Comment (btlogy) [reply mailbox/requirements.in]: Can't we use the git hash to access the archive (w/o relying on git)?
Are you suggesting there could be some differences between sources cloned using Git CLI and those "mirrored" with plain HTTP(s)?

I was under the impression that `tar` and `zip` archives were part of Git itself (e.g.: `git archive`)!

Comment (meejah) [reply mailbox/requirements.in]: > I was under the impression that tar and zip archives were part of Git itself (e.g.: git archive)!

I believe that's correct, but the output isn't stable. So if you have a process that checks hashes, it'll fail sometimes. (Certainly there are ways besides "use the github API" to get the sources of course, I was just highlighting that that particular method has a potential problem).

I believe this discusses the problem: https://github.blog/2023-02-21-update-on-the-future-stability-of-source-code-archives-and-hashes/

**Review by btlogy — APPROVED** [review]
This seem to improve the status quo: now the dependencies are pinned and can be easily parsed (hopefully by dependabot and/or other scan tools).

Thank you for fixing this.

File: mailbox/requirements.in (hunk 1 above)
Comment (meejah) [reply mailbox/requirements.in]: I don't believe that `pip freeze` incantation is doing anything to check the contents...? (e.g. like `--require-hahes` in `pip` might)

**Comment by btlogy** [comment]
As for the upstream dependencies not (yet) been pinned, it could be improved (I can open an issue if that would help).
But this change makes our image more reproducible regardless.
If upstream would pin those deps later, we might be able to skip the `pip-compile` I guess.

**Comment by meejah** [comment]
The `requirements.txt` could still benefit from containing hashes, allowing one to use `--require-hashes` upon install.

(Also, as I said above, if this effort was directed upstream in such a way that releases were done like this, more people would benefit and the actual upstream authors would hopefully take a look at the dependencies themselves)

**Comment by wuan** [comment]
Used hashes with the Git repo which led to errors like 
```
ERROR: Can't verify hashes for these requirements because we don't have a way to hash version control repositories:
```
We can try again, as we are back to the tarball now.

**Comment by wuan** [comment]
Seems to work. Now we might get errors with archives having a different hash in the future, but that should be easy to fix.

**Comment by btlogy** [comment]
> Seems to work. Now we might get errors with archives having a different hash in the future, but that should be easy to fix.

Thank you.

At least, this should bring us a step closer to have an accurate view over the known vulnerabilities in our images.
I'll try to see how we can get there in:
* #37.

**Pull request merged**